### PR TITLE
Indexer sync

### LINKED ISSUE #557: Don't use not synced indexers
Check that indexer reports current head that is reported by nodes.

* Don't use broken indexers in restore
* Don't use broken indexers in new block scan
* Display status that we don't have any operational indexers in history/balance page

### PR DESCRIPTION
**Pull request by LeMarwin**
close #557

### CODE REVIEW

File: wallet/src/Ergvein/Wallet/Monad/Front.hs [hunk 1]
```diff
@@ -231,3 +235,52 @@ setFiltersSync :: MonadFrontAuth t m => Currency -> Bool -> m ()
...
+  let mh = M.lookup cur heights
+  fmap catMaybes $ flip traverse (M.elems conns) $ \con -> do
+    isUp <- sampleDyn $ indexConIsUp con
+    if not isUp then pure Nothing else do
+      mh' <- fmap (M.lookup cur) $ sampleDyn $ indexerConHeight con
+      pure $ case (mh, fromIntegral <$> mh') of
+        (Just h, Just h') -> if h == h' || h - h' == 1 then Just con else Nothing
+        _ -> Nothing
+
+randomElem :: MonadIO m => [a] -> m (Maybe a)
```
Comment (NCrashed) [inline wallet/src/Ergvein/Wallet/Monad/Front.hs:281]: Duplication of https://github.com/hexresearch/ergvein/blob/d30370786dc0f83b4566133acb313380a9bfe505/wallet/src/Ergvein/Wallet/Monad/Client.hs#L214 ?

File: wallet/src/Ergvein/Wallet/Util.hs [hunk 2]
```diff
@@ -27,6 +29,14 @@ widgetHoldDyn maD = do
   ma <- sample . current $ maD
   widgetHold ma $ updated maD
 
+-- | Same as `widgetHold` but tailored for widgets returning events
+widgetHoldE :: forall t m a . (DomBuilder t m, MonadHold t m) => m (Event t a) -> Event t (m (Event t a)) -> m (Event t a)
+widgetHoldE m0 ma = fmap switchDyn $ widgetHold m0 ma
```
Comment (NCrashed) [inline wallet/src/Ergvein/Wallet/Util.hs:34]: Add {-# INLINEABLE widgetHoldE #-}

File: wallet/src/Ergvein/Wallet/Util.hs [hunk 3]
```diff
@@ -27,6 +29,14 @@ widgetHoldDyn maD = do
...
+
+-- | Same as `widgetHoldDyn` but tailored for widgets returning events
+widgetHoldDynE :: forall t m a . (DomBuilder t m, MonadHold t m) => Dynamic t (m (Event t a)) -> m (Event t a)
+widgetHoldDynE = fmap switchDyn . widgetHoldDyn
```
Comment (NCrashed) [inline wallet/src/Ergvein/Wallet/Util.hs:38]: Add {-# INLINEABLE widgetHoldDynE #-}

File: wallet/src/Ergvein/Wallet/Util.hs [hunk 4]
```diff
@@ -79,4 +89,6 @@ currencyToCurrencyCode c = case c of
 currencyCodeToCurrency :: CurrencyCode -> ETC.Currency
 currencyCodeToCurrency c = case c of
   BTC -> ETC.BTC
+  TBTC -> ETC.BTC
   ERGO -> ETC.ERGO
+  TERGO -> ETC.ERGO
```
Comment (NCrashed) [inline wallet/src/Ergvein/Wallet/Util.hs:94]: Partial function, what if indexer send code for DASH?

**Review by NCrashed — CHANGES_REQUESTED** [review]

File: wallet/src/Ergvein/Wallet/Monad/Front.hs (hunk 1 above)
Comment (LeMarwin) [reply wallet/src/Ergvein/Wallet/Monad/Front.hs:281]: No. randomElem was moved from Ergvein.Wallet.Monad.Client to Ergvein.Wallet.Monad.Front

requestRandomIndexer now requires MonadFront, not MonadIndexClient

File: wallet/src/Ergvein/Wallet/Util.hs (hunk 4 above)
Comment (LeMarwin) [reply wallet/src/Ergvein/Wallet/Util.hs:97]: Added error message

**Review by NCrashed — APPROVED** [review]

File: wallet/src/Ergvein/Wallet/Monad/Front.hs [hunk 5]
```diff
@@ -231,3 +235,52 @@ setFiltersSync :: MonadFrontAuth t m => Currency -> Bool -> m ()
...
+requestRandomIndexer reqE = mdo
+  let actE = leftmost [Just <$> reqE, Nothing <$ sentE]
+  sentE <- widgetHoldE (pure never) $ ffor actE $ \case
+    Nothing -> pure never
+    Just (cur, req) -> requester cur req
+  pure sentE
+
+requester :: MonadFront t m => Currency -> Message -> m (Event t (SockAddr, Message))
+requester cur req = mdo
+  connsD  <- externalRefDynamic =<< getActiveConnsRef
```
Comment (NCrashed) [inline wallet/src/Ergvein/Wallet/Monad/Front.hs:249]: Redundant binding

**Review by NCrashed — CHANGES_REQUESTED** [review]

**Pull request merged**